src/avl_tree.jl

```diff
@@ -0,0 +1,259 @@
+        end
+        
+        node.height = fix_height(node)
+        balance = get_balance(node)
+        
+        if balance > 1
+            if key < node.leftChild.data
+                return right_rotate(node)
+            else
+                node.leftChild = left_rotate(node.leftChild)
+                return right_rotate(node)
+            end
+        end
+
+        if balance < -1
+            if key > node.rightChild.data
+                return left_rotate(node)
+            else
+                node.rightChild = right_rotate(node.rightChild)
+                return left_rotate(node)
+            end
+        end
+
+        return node
+    end
+
+    haskey(tree, d) && return tree
+
+    tree.root = insert_node(tree.root, d)
+    tree.count += 1
+    return tree
+end
+
+"""
+    push!(tree, key)
+
+Inserts `key` in the `tree` if it is not present.
+"""
+function Base.push!(tree::AVLTree{K}, key0) where K
+    key = convert(K, key0)
+    insert!(tree, key)
+end
+
+function Base.delete!(tree::AVLTree{K}, d::K) where K
+
+    function delete_node(node::Union{AVLTreeNode, Nothing}, key)
+        if node == nothing || node.data == nothing
+            return nothing
+        elseif key < node.data
+            node.leftChild = delete_node(node.leftChild, key)
+        elseif key > node.data
+            node.rightChild = delete_node(node.rightChild, key)
+        else
+            if node.leftChild == nothing
+                temp = node.rightChild
+                node = nothing
+                return temp
+            elseif node.rightChild == nothing
+                temp = node.leftChild
+                node = nothing
+                return temp
+            else 
+                temp = minimum_node(node.rightChild)
+                node.data = temp.data
+                node.rightChild = delete_node(node.rightChild, temp.data)
+            end 
+        end
+
+        node.height = fix_height(node)
+        balance = get_balance(node)
+
+        if balance > 1
+            if get_balance(node.leftChild) >= 0
+                return right_rotate(node)
+            else
+                node.leftChild = left_rotate(node.leftChild)
+                return right_rotate(node)
+            end
+        end
+
+        if balance < -1
+            if get_balance(node.rightChild) <= 0
+                return left_rotate(node)
+            else
+                node.rightChild = right_rotate(node.rightChild)
+                return left_rotate(node)
+            end
+        end 
+
+        return node
+    end
+
+    # if the key is not in the tree, do nothing and return the tree
+    !haskey(tree, d) && return tree
+    
+    # if the key is present, delete it from the tree
+    tree.root = delete_node(tree.root, d)
+    tree.count -= 1
+    return tree
+end
+
+"""
+    getindex(tree, ind)
+
+Gets the key present at index `ind` of the tree. Indexing is done in increasing order of key.
+"""
+getindex(tree, ind)
```

Comment:
Do we need this docstring?
Or is the docstring that is attached to the fully generic `Base.getindex` clear enough?

This one in parocular is too generically typed I think and will cause problems like:
https://github.com/JuliaCollections/DataStructures.jl/issues/663

Reply:
Removed !!